src/frontend/src/optimizer/plan_node/logical_union.rs

```diff
@@ -107,7 +107,9 @@ impl ToBatch for LogicalUnion {
         let new_inputs: Result<Vec<_>> =
             self.inputs().iter().map(|input| input.to_batch()).collect();
         let new_logical = Self::new(true, new_inputs?);
-        // convert union to union all + agg
+        // We still need to handle !all even if we already have `UnionToDistinctRule`, because it
+        // can be generated by index selection which is an optimization during the `to_batch`.
```

Comment:
Do we have a test covering this case?

Reply:
Yes, we have index selection test. `index_selection.yaml` and `index.slt.part`